### PR TITLE
JNG-4577 tranfrer object auto mapping

### PR DESCRIPTION
**Pull request by robertcsakany**
<table>
<td>
  <a href="https://blackbelt.atlassian.net/browse/JNG-4577" title="JNG-4577" target="_blank">StoryJNG-4577</a>  Transform JSL Service layer to PSM
  </td></table>
  <br />

### CODE REVIEW

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+///////////////////////////////////////////////////////////////////////////
+// Derived relation
+///////////////////////////////////////////////////////////////////////////
+
+@abstract
+rule AbstractCreateTransferObjectDerivedRelationForAutoMappedTransferObjectType
+    transform s : JSL!EntityDerivedDeclaration
+    to t : JUDOPSM!TransferObjectRelation
+    {
+      guard: generateDefaultTransferObject and s.referenceType.isKindOf(JSL!EntityDeclaration)
```
Comment (noherczeg) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl]: Wrong guard

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    to t: JUDOPSM!TransferObjectRelation
+    extends AbstractCreateTransferObjectDerivedRelationForAutoMappedTransferObjectType {
+        guard: JSL!TransferDeclaration.all().exists(t | t.automap and t.map.entity == s.eContainer) and s.referenceType.isKindOf(JSL!EntityDeclaration)
+
+        t.setId("(jsl/" + s.getId() + ")/CreateTransferObjectDerivedRelationForAutoMappedTransferObjectType");
+        t.cardinality = s.equivalentWithPrefix("CreateCardinalityForDerivedDeclaration", t.getId());
+
+        s.eContainer.equivalent("CreateAutoMappedTransferObjectType").relations.add(t);
+
+        log.debug("Created TransferObjectRelation for AutoMappedTransferObjectType: [" + t.name + "] into [" + t.eContainer.name + "]");
```
Comment (noherczeg) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl]: derived

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+rule CloneTransferObjectDerivedRelationForAutoMappedTransferObjectType
+    transform s: JSL!EntityDerivedDeclaration
+    to t: JUDOPSM!TransferObjectRelation 
+    extends AbstractCreateTransferObjectDerivedRelationForAutoMappedTransferObjectType {
+    
+        t.setId("(jsl/" + s.getId() + ")/CloneTransferObjectDerivedRelationForAutoMappedTransferObjectType");
+        t.cardinality = s.equivalentWithPrefix("CreateCardinalityForDerivedDeclaration", t.getId());
+        
+        
+        log.debug("Clone TransferObjectRelation for AutoMappedTransferObjectType: [" + t.name + "]");
```
Comment (noherczeg) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl]: derived

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferAttribute.etl [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+@abstract
+rule AbstractCreateTransferAttributeForTransferFieldDeclaration
+    transform s: JSL!TransferFieldDeclaration
+    to t: JUDOPSM!TransferAttribute {
+        guard: s.referenceType.isKindOf(JSL!PrimitiveDeclaration)
+        t.required = s.isRequired;
+        t.name = s.name;
+        t.dataType = s.referenceType.getPSMEquivalent();
+
+        //if (s.defaultExpression.isDefined()) {
```
Comment (noherczeg) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferAttribute.etl]: why comment this?

**Review by noherczeg — CHANGES_REQUESTED** [review]

**Review by noherczeg — APPROVED** [review]

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferAttribute.etl [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// Create default transfer object attribute and add it to default transfer object
+@greedy
+rule CreateTransferAttributeForAutoMappedTransferObjectType
+    transform s: JSL!EntityFieldDeclaration
+    to t: JUDOPSM!TransferAttribute
+    extends AbstractCreateTransferAttributeForAutoMappedTransferObjectType {
+        guard: JSL!TransferDeclaration.all().exists(t | t.automap and t.map.entity == s.eContainer) and s.referenceType.isKindOf(JSL!PrimitiveDeclaration)
+        t.setId("(jsl/" + s.getId() + ")/CreateTransferAttributeForAutoTransferObjectType");
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferAttribute.etl]: > `t.setId("(jsl/" + s.getId() + ")/CreateTransferAttributeForAutoTransferObjectType");`

CreateTransferAttributeForAutoMappedTransferObjectType?

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferAttribute.etl [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+rule CreateDefaultTransferAttributeForAutoMappedTransferObjectType
+    transform s: JSL!EntityDerivedDeclaration    
+    to t: JUDOPSM!TransferAttribute
+    extends AbstractCreateTransferDefaultValueAttributeForAutoMappedTransferObjectType {
+        guard: JSL!TransferDeclaration.all().exists(t | t.automap and t.map.entity == s.eContainer) and s.referenceType.isKindOf(JSL!PrimitiveDeclaration)
+        t.setId("(jsl/" + s.getId() + ")/CreateDerivedTransferAttributeForAutoMappedTransferObjectType");
+
+        s.eContainer.equivalent("CreateEntityAutoMappedTransferObjectType").attributes.add(t);
+        log.debug("Created TransferAttribute (Derived) for AutoMappedTransferObjectType: [" + t.name + "] into [" + t.eContainer.name + "]");
+}
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferAttribute.etl]: Both **default** and **derived** terms used in this case. Is it intentional or a typo?

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferObjectType.etl [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+						if (im.defaultExpression.isDefined()) {
+    						t.attributes.add(im.defaultExpression.equivalentWithPrefix("CloneTransferDefaultValueAttributeForAutoMappedTransferObjectType", t.getId()));
+						}
+					// Embedded relations
+	                } else if (im.referenceType.isKindOf(JSL!EntityDeclaration)) {
+						t.relations.add(im.equivalentWithPrefix("CloneTransferObjectEmbeddedRelationForAutoMappedTransferObjectType", t.getId()));	                
+	                }
+                // Derives
+		        } else if (im.isKindOf(JSL!EntityDerivedDeclaration)) {
+					// Primitive derives
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferObjectType.etl]: > Derives

Derived

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+@lazy
+@greedy
+rule CloneTransferObjectEmbeddedRelationForAutoMappedTransferObjectType
+    transform s: JSL!EntityFieldDeclaration
+    to t: JUDOPSM!TransferObjectRelation 
+    extends AbstractCreateTransferObjectEmbeddedRelationForAutoMappedTransferObjectType {
+    
+        t.setId("(jsl/" + s.getId() + ")/CloneTransferObjectEmbeddedRelationForDefaultTransferObjectType");
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/entityDeclarationAutoMappedTransferRelation.etl]: > `t.setId("(jsl/" + s.getId() + ")/CloneTransferObjectEmbeddedRelationForDefaultTransferObjectType");`

CloneTransferObjectEmbeddedRelationForAutoMappedTransferObjectType?

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferAttribute.etl [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      t.setId("(jsl/" + s.getId() + ")/CreateDerivedTransferAttribute");      
+      
+      if (s.eContainer.map.isDefined()) {
+         t.binding = s.reads.equivalent("CreateReadsDataPropertyForMappedTransferObjectTranferFieldDeclaration");
+      } else {
+         t.binding = s.reads.equivalent("CreateReadsStaticDataForUnmappedTransferObjectTranferFieldDeclaration");
+      }
+      
+      s.eContainer.getPSMEquivalent().attributes.add(t);
+      log.debug("Created TransferAttribute (Field) for TransferFieldDeclaration: [" + t.name + "] into [" + t.eContainer.name + "]");
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferAttribute.etl]: > TransferAttribute (Field)

TransferAttribute (Derived)?

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    to t: JUDOPSM!TransferObjectRelation {
+        guard: s.referenceType.isKindOf(JSL!TransferDeclaration)
+
+        t.name = s.name;
+        t.target = s.referenceType.getPSMEquivalent();
+
+        // TODO: 
+        // t.embeddedCreate = not s.eContainer.isAbstract;
+        // t.embeddedUpdate = not s.eContainer.isAbstract;
+        // t.embeddedDelete = not s.eContainer.isAbstract;
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl:18]: WIP or ticket?

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+@greedy
+rule CreateDerivedTransferObjectEmbeddedRelationForTransferFieldDeclaration
+    transform s : JSL!TransferFieldDeclaration
+    to t : JUDOPSM!TransferObjectRelation
+    extends AbstractCreateTransferObjectEmbeddedRelationForTransferFieldDeclaration {
+      guard: s.referenceType.isKindOf(JSL!TransferDeclaration) and s.reads.isDefined() and not s.eContainer.automap
+
+      t.setId("(jsl/" + s.getId() + ")/CreateTransferObjectEmbeddedRelationForMappedTransferFieldDeclaration");
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl]: > t.setId("(jsl/" + s.getId() + ")/CreateTransferObjectEmbeddedRelationForMappedTransferFieldDeclaration");

CreateDerivedTransferObjectEmbeddedRelationForTransferFieldDeclaration?

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+@greedy
+rule CreateMappedTransferObjectEmbeddedRelationForTransferFieldDeclaration
+    transform s : JSL!TransferFieldDeclaration
+    to t : JUDOPSM!TransferObjectRelation
+    extends AbstractCreateTransferObjectEmbeddedRelationForTransferFieldDeclaration {
+      guard: s.referenceType.isKindOf(JSL!TransferDeclaration) and s.maps.isDefined() and not s.eContainer.automap
+
+      t.setId("(jsl/" + s.getId() + ")/CreateTransferObjectEmbeddedRelationForMappedTransferFieldDeclaration");
```
Comment (bencegelei) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl]: CreateMappedTransferObjectEmbeddedRelationForTransferFieldDeclaration or CreateTransferObjectEmbeddedRelationForMappedTransferFieldDeclaration

**Review by bencegelei — CHANGES_REQUESTED** [review]

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl (hunk 10 above)
Comment (robertcsakany) [reply judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationTransferRelation.etl:18]: WIP

**Review by bencegelei — APPROVED** [review]

**Pull request merged**